socs/agents/pysmurf_controller/smurf_subprocess_util.py

```diff
@@ -0,0 +1,290 @@
+    func_name: str
+    "name of function to run"
+
+    args: List = field(default_factory=list)
+    "Args to pass to function"
+
+    kwargs: Dict = field(default_factory=dict)
+    "Kwargs to pass to function"
+
+    slot: Optional[int] = None
+    "Slot to setup smurf control for. If None, uses the slot from the environment variable 'SLOT'"
+
+    def __post_init__(self):
+        if self.func_name not in func_map:
+            raise ValueError(f"Invalid function name: {self.func_name}")
+
+
+@dataclass
+class RunResult:
+    success: bool
+    "True if function returned succesfully"
+
+    return_val: any = None
+    "Return value from function"
+
+    traceback: Optional[str] = None
+    "Traceback if function raised an exception"
+
+
+class FuncProtocol(protocol.ProcessProtocol):
+    """
+    Process protocol for running one of the above functions in a subprocess.
+    After the subprocess is started, the RunCfg object is encoded and sent to
+    the subprocess through stdin. Once the function is finished, the result is
+    encoded and passed back through FD 3.
+    """
+
+    def __init__(self, cfg: RunCfg):
+        self.cfg = cfg
+        self.result = None
+
+    def connectionMade(self):
+        data = json.dumps(asdict(self.cfg)).encode()
+        self.transport.write(data)
+        self.transport.closeStdin()
+
+    def childDataReceived(self, childFD, data):
+        if childFD in [1, 2]:  # stdout or stderr:
+            print(data.decode())
+
+        if childFD == 3:
+            self.result = json.loads(data.decode())
+
+    def processExited(self, status):
+        if self.result is None:
+            raise RuntimeError("No result received from child process")
+
+        self.deferred.callback(self.result)
+
+
+@defer.inlineCallbacks
+def _run_func_in_subprocess_reactor(cfg: RunCfg) -> RunResult:
+    """Helper function for run_func_in_subprocess, that can assume reactor context"""
+    prot = FuncProtocol(cfg)
+    prot.deferred = defer.Deferred()
+    childFDs = {0: 'w', 1: 'r', 2: 'r', 3: 'r'}  # Regular FDs plus 3 for sending results back
+    env = os.environ.copy()
+    if cfg.slot is not None:
+        env['SLOT'] = str(cfg.slot)
+    reactor.spawnProcess(
+        prot, sys.executable, [sys.executable, '-u', __file__], childFDs=childFDs,
+        env=env
+    )
+    result = yield prot.deferred
+    return RunResult(**result)
+
+
+def run_func_in_subprocess(cfg: RunCfg) -> RunResult:
+    """
+    This function takes a RunCfg object, and runs the specified function in a
+    subprocess. The result is returned as a RunResult object. This function
+    must be run in the reactor thread.
+
+    Args
+    -----
+    cfg: RunCfg
+        Configuration object to specify the function to run, and the arguments.
+    """
+    if not in_reactor_context():
+        return threads.blockingCallFromThread(
+            reactor, _run_func_in_subprocess_reactor, cfg)
+    else:
+        return _run_func_in_subprocess_reactor(cfg)
```

Comment:
Cool... I believe that if you _are_ in reactor context, then you can't block -- i.e you need to return a Deferred here, which would only be possible if `_run_func_in_subprocess_reactor(cfg)` returned `prot.deferred` or some wrapped thing that applies the Runresult.

Reply:
Good point... I changed to `run_func_in_subprocess_from_thread` and had it only handle the from-thread case